bookdown/07-special-cluster.Rmd

```diff
@@ -0,0 +1,109 @@
+  * `target` field is empty and will return `character(0)` if accessed anyway.
+
+
+### Train and Predict
+Clustering learners provide both `train` and `predict` methods.
+The analysis typically consists of building clusters using all available data.
+Nevertheless, some learners can assign new observations to existing groups with `predict`.
+However, prediction does not always make sense, e.g., for hierarchical clustering.
+Moreover, some learners estimate the probability of each observation belonging to a given cluster.
+`predict_types` field gives a list of prediction types for each learner.
+
+After training, the `model` field stores a learner's model that looks different for each learner depending on the underlying library.
+`predict` returns a `PredictionClust` object that gives a simplified view of the learned model.
+If the data given to the `predict` method is the same as the one on which the learner was trained, `predict` simply returns cluster assignments for the "training" observations.
+On the other hand, if the test set contains new data, `predict` will estimate cluster assignments for that data set (some learners do not do that, so read the documentation first).
+
+In the following example, a `r ref("mlr_learners_clust.kmeans", text = "$k$-means learner")` is applied on the `r ref("mlr_tasks_usarrests", text = "US arrest data set")`.
+The class labels are predicted and the contribution of the task features to assignment of the respective class are visualized.
+```{r 07-special-cluster-002, message=FALSE, warning=FALSE}
+library("mlr3")
+library("mlr3cluster")
+library("mlr3viz")
+
+# create an example task
+task = tsk("usarrests")
+print(task)
+autoplot(task)
+
+# create a k-means learner
+learner = lrn("clust.kmeans")
+
+# assigning each observation to one of the two clusters (default in clust.kmeans)
+learner$train(task)
+learner$model
+
+# make "predictions" for the same data
+prediction = learner$predict(task)
+autoplot(prediction, task)
+```
+
+
+### Measures
+To assess the quality of clustering, there are a few built-in evaluation metrics available.
+One of them is silhouette quality index that measures how well each point belongs to its assigned cluster versus neighboring cluster.
+Silhouette values are in $[-1, 1]$ range.
+
+Points with silhouette closer to:
+
+   * 1 are well clustered
+   * 0 lie between two clusters
+   * -1 likely placed in the wrong cluster
+
+Using silhouette plots, one can visually assess the quality of the analysis and choose the most optimal number of clusters for the data.
+The red dotted line shows the mean silhouette value and each bar represents a data point.
+If most points in each cluster have an index higher than the mean silhouette, the number of clusters is chosen well.
+
+```{r 07-special-cluster-003, message=FALSE, warning=FALSE}
+# silhouette plot allows to visually inspect the quality of clustering
+autoplot(prediction, task, type = "sil")
```

Comment:
Presumably there's a way to get an average score as well, so you can assess this without a plot?